[mlir][transform] Check for invalidated iterators on payload IR mappings

matthias-springer opened the pull request:
Add extra error checking (in debug mode) to detect cases where an iterator on "direct" payload IR mappings is invalidated (due to elements being removed). Such errors are hard to debug: they are often non-deterministic; sometimes the program crashes, sometimes it produces wrong results. Even when it crashes, the stack trace often points to completely unrelated code locations.

Store a timestamp with each "direct" mapping. The timestamp is increased whenever an operation is performed that invaldiates an iterator on that mapping. A debug iterator is added that checks the timestamp as payload IR is enumerated.

matthias-springer commented on mlir/test/lib/Dialect/Transform/TestTransformDialectExtension.cpp:
```diff
@@ -360,8 +360,9 @@ DiagnosedSilenceableFailure mlir::test::TestRemoveTestExtensionOp::apply(
 DiagnosedSilenceableFailure mlir::test::TestReversePayloadOpsOp::apply(
     transform::TransformRewriter &rewriter,
     transform::TransformResults &results, transform::TransformState &state) {
-  auto payloadOps = state.getPayloadOps(getTarget());
```
This change is not needed.

ftynse commented on mlir/include/mlir/Dialect/Transform/IR/TransformInterfaces.h:
```diff
@@ -178,6 +184,11 @@ class TransformState {
     ValueMapping values;
     ValueMapping reverseValues;
+
+#ifndef NDEBUG
```
Code that may change the size of the object should be guarded by `LLVM_ENABLE_ABI_BREAKING_CHECKS`.

ftynse reviewed (approved):
LGTM with the correct macro predicate

matthias-springer committed:
[mlir][transform] Check for invalidated iterators on payload IR mappings

Add extra error checking (in debug mode) to detect cases where an iterator on "direct" payload IR mappings is invalidated (due to elements being removed). Such errors are hard to debug: they are often non-deterministic; sometimes the program crashes, sometimes it produces wrong results. Even when it crashes, the stack trace often points to completely unrelated code locations.

Store a timestamp with each "direct" mapping. The timestamp is increased whenever an operation is performed that invaldiates an iterator on that mapping. A debug iterator is added that checks the timestamp before dereferencing or incrementing.
